README
trytond_bank_es
===============

The bank_es module of the Tryton application platform.

Installing
----------

See INSTALL

Support
-------

For more information or if you encounter any problems with this module,
please contact the programmers at

  Zikzakmedia
  --------------
  website: http://www.zikzakmedia.com/
  email: zikzak@zikzakmedia.com

If you encounter any problems with Tryton, please don't hesitate to ask
questions on the Tryton bug tracker, mailing list, wiki or IRC channel:

  http://bugs.tryton.org/
  http://groups.tryton.org/
  http://wiki.tryton.org/
  irc://irc.freenode.net/tryton

License
-------

See LICENSE

Copyright
---------

See COPYRIGHT


For more information please visit the Tryton web site:

  http://www.tryton.org/


CHANGELOG
Version 5.5.0 - 2019-11-14
Version 5.4.0 - 2019-11-14
Version 5.2.0 - 2019-05-07
Version 5.3.0 - 2019-05-06
Version 4.8.0 - 2018-04-25

Version 4.2.0 - 2016-11-28
* Bug fixes (see mercurial logs for details)

Version 4.0.0 - 2016-05-03
Version 3.4.0 - 2014-11-03

Version 3.2.0 - 2014-05-22

Version 3.0.0 - 2013-10-25
* Initial Release


tests/__init__.py
# This file is part bank_es module for Tryton.
# The COPYRIGHT file at the top level of this repository contains the full
# copyright notices and license terms.
try:
    from trytond.modules.bank_es.tests.test_bank_es import suite
except ImportError:
    from .test_bank_es import suite

__all__ = ['suite']


tests/test_bank_es.py
# This file is part of the bank_es module for Tryton.
# The COPYRIGHT file at the top level of this repository contains the full
# copyright notices and license terms.
import unittest
import trytond.tests.test_tryton
from trytond.tests.test_tryton import ModuleTestCase, with_transaction
from trytond.error import UserError
from trytond.pool import Pool


def import_banks():
    pool = Pool()
    LoadBank = pool.get('load.banks', type='wizard')
    Country = pool.get('country.country')
    # We must create a country as country data is not loaded on tests
    if not Country.search([('code', '=', 'ES')]):
        country = Country(code='ES', name='Spain')
        country.save()
    session_id, _, _ = LoadBank.create()
    load = LoadBank(session_id)
    load.transition_accept()


class BankEsTestCase(ModuleTestCase):
    'Test Bank Es module'
    module = 'bank_es'

    @with_transaction()
    def test_import_banks(self):
        'Import Spanish Banks'
        pool = Pool()
        Bank = pool.get('bank')
        self.assertEqual(Bank.search([], count=True), 0)
        import_banks()
        banks = Bank.search([], count=True)
        self.assertGreater(banks, 0)
        # Update should not duplicate
        import_banks()
        self.assertEqual(banks, Bank.search([], count=True))

    @with_transaction()
    def test_spanish_ccc_validation(self):
        'Spanish CCC Number'
        pool = Pool()
        Party = pool.get('party.party')
        Bank = pool.get('bank')
        Account = pool.get('bank.account')

        party = Party(name='Test')
        party.save()
        bank = Bank(party=party, bank_code='BNK')
        bank.save()
        with self.assertRaises(UserError):
            Account.create([{
                        'bank': bank.id,
                        'numbers': [('create', [{
                                        'type': 'es_ccc',
                                        'number': '1234-1234-00 1234567890',
                                        }])],
                        }])

    @with_transaction()
    def test_spanish_ccc_number(self):
        'Spanish CCC Number'
        pool = Pool()
        Party = pool.get('party.party')
        Bank = pool.get('bank')
        Account = pool.get('bank.account')

        party = Party(name='Test')
        party.save()
        bank = Bank(party=party, bank_code='BNK')
        bank.save()
        account, = Account.create([{
                    'bank': bank.id,
                    'numbers': [('create', [{
                                    'type': 'es_ccc',
                                    'number': '21000418400200051331',
                                    }])],
                    }])

        iban_number, es_number = account.numbers
        self.assertEqual(es_number.number_compact, '21000418400200051331')
        self.assertEqual(es_number.number, '2100 0418 40 02000 51331')
        self.assertEqual(iban_number.number_compact,
            'ES4021000418400200051331')

        account, = Account.create([{
                    'bank': bank.id,
                    'numbers': [('create', [{
                                    'type': 'other',
                                    'number': '21000418400200051331',
                                    }])],
                    }])

        # Test not created until es_ccc
        number, = account.numbers
        number.type = 'es_ccc'
        number.save()
        iban_number, es_number = account.numbers
        # Test iban not duplicated
        account, = Account.create([{
                    'bank': bank.id,
                    'numbers': [('create', [{
                                    'type': 'es_ccc',
                                    'number': '21000418400200051331',
                                    }, {
                                    'type': 'iban',
                                    'number': 'ES4021000418400200051331',
                                    }])],
                    }])

        iban_number, es_number = account.numbers


def suite():
    suite = trytond.tests.test_tryton.suite()
    suite.addTests(unittest.TestLoader().loadTestsFromTestCase(
        BankEsTestCase))
    return suite


download_bce_data.py
import csv
import tempfile

import urllib.request, urllib.parse, urllib.error
from xlrd import open_workbook

urlfile = urllib.request.URLopener()
with tempfile.NamedTemporaryFile() as banks:
    urlfile.retrieve('http://www.bde.es/f/webbde/SGE/regis/ficheros/es/'
        'REGBANESP_CONESTAB_A.XLS', banks.name)
    book = open_workbook(banks.name)
    # We filter by 'bancos, cajas de ahorro, cooperativas de credito, entidades
    # de credito comunitarias y extracomunitarias'
    bank_types = ['BP', 'CA', 'CC', 'CO', 'EFC', 'OR', 'SECC', 'SECE']
    bic_codes = {}

    with open('bic.csv', 'rt') as bic:
        reader = csv.reader(bic, delimiter=',')
        for code in reader:
            bic_codes[code[0]] = code[2]
    # The bank info is in the first sheet
    sheet = book.sheet_by_index(0)
    headers = [sheet.cell(0, col_index).value
        for col_index in range(sheet.ncols)]
    headers.append('BIC')
    with open('bank.csv', 'wt') as f:
        writer = csv.DictWriter(f, fieldnames=headers, extrasaction='ignore',
                quoting=csv.QUOTE_NONNUMERIC, lineterminator="\n",
                delimiter=',', quotechar='"')
        writer.writeheader()
        for row_index in range(1, sheet.nrows):
                d = {headers[col_index]: sheet.cell(row_index, col_index).value
                    for col_index in range(sheet.ncols)}
                if d['COD_TIPO'] in bank_types and d['FCHBAJA'] == '':
                    for value in list(d):
                        # Encode all the values to prevent an import crash
                        d[value] = d[value].encode('utf-8')
                        d['BIC'] = None
                        if d['COD_BE'] in bic_codes:
                            d['BIC'] = bic_codes[d['COD_BE']]
                    writer.writerow(d)


bank.py
# This file is part of bank_es module for Tryton.
# The COPYRIGHT file at the top level of this repository contains
# the full copyright notices and license terms.
from csv import reader
import os

from trytond.model import ModelView, fields
from trytond.pool import Pool, PoolMeta
from trytond.pyson import Not, Eval, Bool
from trytond.wizard import Button, StateView, Wizard, StateTransition
from trytond.transaction import Transaction

__all__ = ['Bank', 'LoadBanksStart', 'LoadBanks']


class Bank(metaclass=PoolMeta):
    __name__ = 'bank'
    bank_code = fields.Char('National Code', select=1,
        states={
            'required': Not(Bool(Eval('bic')))
            }, depends=['bic'])

    @classmethod
    def __setup__(cls):
        super(Bank, cls).__setup__()
        cls.party.context.update({
                'active_test': False,
                })

    @classmethod
    def search_rec_name(cls, name, clause):
        domain = super(Bank, cls).search_rec_name(name, clause)
        if clause[1].startswith('!') or clause[1].startswith('not '):
            bool_op = 'AND'
        else:
            bool_op = 'OR'
        return [bool_op,
            domain,
            ('party',) + tuple(clause[1:]),
            ('bic',) + tuple(clause[1:]),
            ('bank_code',) + tuple(clause[1:]),
            ]


class LoadBanksStart(ModelView):
    '''Load Banks Start'''
    __name__ = 'load.banks.start'


class LoadBanks(Wizard):
    '''Load Banks'''
    __name__ = "load.banks"

    start = StateView('load.banks.start',
        'bank_es.load_banks_start_view_form', [
            Button('Cancel', 'end', 'tryton-cancel'),
            Button('Accept', 'accept', 'tryton-ok', default=True),
            ])
    accept = StateTransition()

    @classmethod
    def __setup__(cls):
        super(LoadBanks, cls).__setup__()
        cls._error_messages.update({
                'error': 'CSV Import Error!',
                'read_error': 'Error reading file: %s.\nError raised: %s.',
                })

    def transition_accept(self):
        def get_subdivision(code):
            subdividion = {
                '01': 'ES-VI',
                '02': 'ES-AB',
                '03': 'ES-A',
                '04': 'ES-AL',
                '05': 'ES-AV',
                '06': 'ES-BA',
                '07': 'ES-IB',
                '08': 'ES-B',
                '09': 'ES-BU',
                '10': 'ES-CC',
                '11': 'ES-CA',
                '12': 'ES-CS',
                '13': 'ES-CR',
                '14': 'ES-CO',
                '15': 'ES-C',
                '16': 'ES-CU',
                '17': 'ES-GI',
                '18': 'ES-GR',
                '19': 'ES-GU',
                '20': 'ES-SS',
                '21': 'ES-H',
                '22': 'ES-HU',
                '23': 'ES-J',
                '24': 'ES-LE',
                '25': 'ES-L',
                '26': 'ES-LO',
                '27': 'ES-LU',
                '28': 'ES-M',
                '29': 'ES-MA',
                '30': 'ES-MU',
                '31': 'ES-NA',
                '32': 'ES-OU',
                '33': 'ES-O',
                '34': 'ES-P',
                '35': 'ES-GC',
                '36': 'ES-PO',
                '37': 'ES-SA',
                '38': 'ES-TF',
                '39': 'ES-S',
                '40': 'ES-SG',
                '41': 'ES-SE',
                '42': 'ES-SO',
                '43': 'ES-T',
                '44': 'ES-TE',
                '45': 'ES-TO',
                '46': 'ES-V',
                '47': 'ES-VA',
                '48': 'ES-BI',
                '49': 'ES-ZA',
                '50': 'ES-Z',
                '51': 'ES-CE',
                '52': 'ES-ML',
                }
            return subdividion.get(code, None)

        pool = Pool()
        Lang = pool.get('ir.lang')
        lang, = Lang.search([('code', '=', 'es')])
        Bank = pool.get('bank')
        Party = pool.get('party.party')
        Identifier = pool.get('party.identifier')
        Address = pool.get('party.address')
        Country = pool.get('country.country')
        country, = Country.search([('code', '=', 'ES')])
        Contact = pool.get('party.contact_mechanism')
        Subdivision = pool.get('country.subdivision')
        transaction = Transaction()

        delimiter = ','
        quotechar = '"'

        def get_rows():
            with open(os.path.join(os.path.dirname(__file__), 'bank.csv'), 'r',
                    encoding='utf-8') as data:
                try:
                    rows = list(reader(data, delimiter=delimiter,
                        quotechar=quotechar))
                except TypeError as e:
                    self.raise_user_error('error',
                        error_description='read_error',
                        error_description_args=('bank.csv', e))
                return rows[1:]
        created_parties = {}
        for row in get_rows():
            if not row:
                continue

            with transaction.set_context(active_test=False):
                parties = Party.search(['OR',
                        ('name', '=', row[4]),
                        ('code', '=', 'BNC' + row[1])
                        ])
            if parties:
                party = parties[0]
            else:
                party = Party()
                party.active = False
                party.name = row[4]
                party.code = 'BNC' + row[1]
                party.addresses = []
                party.lang = lang
                party.addresses = []
                party.identifiers = []
                party.contact_mechanisms = []
            if row[6]:
                codes = set([c.code for c in party.identifiers])
                vat_code = 'ES%s' % row[6]
                if vat_code not in codes:
                    identifier = Identifier()
                    identifier.type = 'eu_vat'
                    identifier.code = vat_code
                    party.identifiers = [identifier] + list(party.identifiers)

            addresses = set([a.name for a in party.addresses])
            if party.name not in addresses:
                address = Address()
                address.active = False
                address.name = party.name
                address.street = row[8]
                address.zip = row[11]
                address.city = row[12]
                address.country = country
                subdivisions = Subdivision.search([
                        ('code', '=', get_subdivision(row[14])),
                        ], limit=1)
                if subdivisions:
                    address.subdivision, = subdivisions
                party.addresses = [address] + list(party.addresses)

            new_mechanisms = []
            current_mechanisms = set([(c.type, c.value)
                    for c in party.contact_mechanisms])
            if row[16] and ('phone', row[16]) not in current_mechanisms:
                contact = Contact()
                contact.type = 'phone'
                contact.value = row[16]
                if row[16].startswith('0'):
                    contact.value = row[16][1:]
                new_mechanisms.append(contact)

            if row[18] and ('fax', row[18]) not in current_mechanisms:
                contact = Contact()
                contact.type = 'fax'
                contact.value = row[18]
                if row[18].startswith('0'):
                    contact.value = row[18][1:]
                new_mechanisms.append(contact)

            if (row[19] and ('website',
                        row[19].lower()) not in current_mechanisms):
                contact = Contact()
                contact.type = 'website'
                contact.value = row[19].lower()
                new_mechanisms.append(contact)
            if new_mechanisms:
                party.contact_mechanisms = (new_mechanisms +
                    list(party.contact_mechanisms))
            created_parties[row[1]] = party
        Party.save(list(created_parties.values()))

        to_save = []
        for row in get_rows():
            if not row:
                continue
            banks = Bank.search([('bank_code', '=', row[1])])
            if not banks:
                bank = Bank(party=None, bank_code=None, bic=None)
            else:
                bank = banks[0]
            party = created_parties[row[1]]
            if bank.party != party:
                bank.party = party
            if bank.bank_code != row[1]:
                bank.bank_code = row[1]
            if bank.bic != row[47]:
                bank.bic = row[47]
            to_save.append(bank)
        Bank.save(to_save)
        return 'end'
